storm-core/src/jvm/org/apache/storm/daemon/supervisor/SyncProcessEvent.java

```diff
@@ -0,0 +1,427 @@
+ * ids, write new "approved workers" to LS 5. create local dir for worker id 5. launch new workers (give worker-id, port, and supervisor-id) 6. wait for workers
+ * launch
+ */
+public class SyncProcessEvent implements Runnable {
+
+    private static Logger LOG = LoggerFactory.getLogger(SyncProcessEvent.class);
+
+    private  LocalState localState;
+    private  SupervisorData supervisorData;
+    public static final ExecutorInfo SYSTEM_EXECUTOR_INFO = new ExecutorInfo(-1, -1);
+
+    private class ProcessExitCallback implements Utils.ExitCodeCallable {
+        private final String logPrefix;
+        private final String workerId;
+
+        public ProcessExitCallback(String logPrefix, String workerId) {
+            this.logPrefix = logPrefix;
+            this.workerId = workerId;
+        }
+
+        @Override
+        public Object call() throws Exception {
+            return null;
+        }
+
+        @Override
+        public Object call(int exitCode) {
+            LOG.info("{} exited with code: {}", logPrefix, exitCode);
+            supervisorData.getDeadWorkers().add(workerId);
+            return null;
+        }
+    }
+
+    public SyncProcessEvent(){
+
+    }
+    public SyncProcessEvent(SupervisorData supervisorData) {
+        init(supervisorData);
+    }
+    
+    public void init(SupervisorData supervisorData){
+        this.supervisorData = supervisorData;
+        this.localState = supervisorData.getLocalState();
+    }
+
+    @Override
+    public void run() {
+        LOG.debug("Syncing processes");
+        try {
+            Map conf = supervisorData.getConf();
+            Map<Integer, LocalAssignment> assignedExecutors = localState.getLocalAssignmentsMap();
+
+            if (assignedExecutors == null) {
+                assignedExecutors = new HashMap<>();
+            }
+            int now = Time.currentTimeSecs();
+
+            Map<String, StateHeartbeat> localWorkerStats = getLocalWorkerStats(supervisorData, assignedExecutors, now);
+
+            Set<String> keeperWorkerIds = new HashSet<>();
+            Set<Integer> keepPorts = new HashSet<>();
+            for (Map.Entry<String, StateHeartbeat> entry : localWorkerStats.entrySet()) {
+                StateHeartbeat stateHeartbeat = entry.getValue();
+                if (stateHeartbeat.getState() == State.VALID) {
+                    keeperWorkerIds.add(entry.getKey());
+                    keepPorts.add(stateHeartbeat.getHeartbeat().get_port());
+                }
+            }
+            Map<Integer, LocalAssignment> reassignExecutors = getReassignExecutors(assignedExecutors, keepPorts);
+            Map<Integer, String> newWorkerIds = new HashMap<>();
+            for (Integer port : reassignExecutors.keySet()) {
+                newWorkerIds.put(port, Utils.uuid());
+            }
+            LOG.debug("Syncing processes");
```

Comment:
minor nit. We are logging this twice here and up above.  It would probably be good to just do it once.